Download functionality for files and pipelines

PopaRares opened the pull request:
# Description
Added download functionality for files and pipelines. Users are now able to right click on any file in the **Files** tab and chose the new **Download file** option, or click the **save icon** on the pipelines page to download an entire pipeline in the form of a zip file.

![file_download](https://github.com/mage-ai/mage-ai/assets/33767023/98a256f1-1694-43e2-bf48-617a0ab9b43e)
*Downloading a file*

![pipeline_download](https://github.com/mage-ai/mage-ai/assets/33767023/a7c50824-742a-4fd4-b00d-153d80ac0014)
*Downloading a pipeline*

When downloading a pipeline, users can chose to preserve its folder structure, or download the files only.

![context_menu](https://github.com/mage-ai/mage-ai/assets/33767023/3ea5d125-a999-45de-96fa-2757c54afbc9)
![comparison](https://github.com/mage-ai/mage-ai/assets/33767023/bf0aa3b1-1de1-475b-8308-f537141e0ad4)
*zip comparison between normal and file-only pipeline download*

## How it works
Each download request involves two separate calls.
1. **Token Generation**:
   - The first call creates a JWT token that contains the file path of the requested resource.
   - In the case of a pipeline download, the token will contain a list of file paths of all related files.
   - The token is sent to the client which generates the dynamic download link used in the second call.

2. **File Retrieval**:
   - The second call takes place when the client programmatically clicks on the generated link.
   - In the case of a pipeline download, all related files get zipped together in a temporary zip file.
   - This action prompts the server to send the requested file using a [Content-Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) call.
   - The temporary zip is automatically deleted after download (in the case of a pipeline download). 

# How Has This Been Tested?
Manually tested from the UI.


# Checklist
- [x] The PR is tagged with proper labels (bug, enhancement, feature, documentation)
- [x] I have performed a self-review of my own code
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

cc:
@hugabora @adelcast

wangxiaoyou1993 commented:
cc: @johnson-mage

johnson-mage commented on mage_ai/frontend/components/FileBrowser/index.tsx:
```diff
@@ -90,6 +90,33 @@ function FileBrowser({
+      onSuccess: (response: any) => onSuccess(
+        response, {
+          callback: () => {
+            const url = response.data.download.uri
```
linting: missing semicolon

johnson-mage commented on mage_ai/frontend/components/FileBrowser/index.tsx:
```diff
@@ -474,6 +501,13 @@ function FileBrowser({
+          label: () => 'Download file',
+          onClick: () => {
+            const fp = getFullPathWithoutRootFolder(selectedFile);
+            downloadFile(encodeURIComponent(fp));
```
Saw a linting error for a missing dependency, so you may need to include `downloadFile` in the useMemo dependency array here: https://github.com/mage-ai/mage-ai/pull/3813/files#diff-5df45198d443d4d572868f9c02c6c2d71a000c912c5a5b254ce42033ae6c930cR568-R581

johnson-mage commented on mage_ai/frontend/pages/pipelines/index.tsx:
```diff
@@ -1377,6 +1432,16 @@ function PipelineListPage() {
+            <Button
+              {...sharedOpenButtonProps}
+              onClick={() => {
+                downloadPipeline({pipelineUUID: uuid});
```
linting: Include space between brackets (i.e. `downloadPipeline({ pipelineUUID: uuid });`)

johnson-mage commented on mage_ai/frontend/pages/pipelines/index.tsx:
```diff
@@ -1204,6 +1239,26 @@ function PipelineListPage() {
+          {
+            label: () => 'Download (without folder structure)',
+            onClick: () => {
+              downloadPipeline({
```
Missing dependency `downloadPipeline` in useCallback dependency array here: https://github.com/mage-ai/mage-ai/pull/3813/files#diff-22d0517fd466848fa8e87718671265bcaecd728c9a6a4407b128c3532a88c6d9R1479-R1494

johnson-mage commented on mage_ai/frontend/pages/pipelines/index.tsx:
```diff
@@ -288,6 +289,40 @@ function PipelineListPage() {
+      pipelineUUID: string;
+      filesOnly?: boolean;
+    }) => 
+      api.downloads.pipelines.useCreate(pipelineUUID)({'download': {'ignore_folder_structure': filesOnly}}),
```
linting: spaces between brackets. May also be more readable to format this over multiple lines like this:
```
api.downloads.pipelines.useCreate(pipelineUUID)(
  { 'download': { 'ignore_folder_structure': filesOnly } },
),
```

PopaRares replied:
Left the curly brackets next to the parentheses; hope this works.

johnson-mage replied:
yup, that works too!

johnson-mage commented on mage_ai/frontend/pages/pipelines/index.tsx:
```diff
@@ -1204,6 +1239,26 @@ function PipelineListPage() {
+                filesOnly: false
+              });
+            },
+            uuid: 'download',
```
Let's differentiate the uuids (e.g. `download_keep_folder_structure`).

johnson-mage commented on mage_ai/frontend/pages/pipelines/index.tsx:
```diff
@@ -1204,6 +1239,26 @@ function PipelineListPage() {
+                filesOnly: true
+              });
+            },
+            uuid: 'download',
```
Let's differentiate the uuids (e.g. `download_without_folder_structure`).

johnson-mage commented:
@wangxiaoyou1993 Can you take a look at the backend code please?

johnson-mage commented:
@PopaRares Can you please test with [user authentication enabled](https://docs.mage.ai/production/authentication/overview#require-user-authentication) and as an "editor" role? You'll need to update the `DownloadPolicy` to avoid some policy errors. For example: 
![image](https://github.com/mage-ai/mage-ai/assets/78053898/1d4c7f8d-7af6-4a86-b244-1078e2d03e90)

PopaRares commented:
> @PopaRares Can you please test with [user authentication enabled](https://docs.mage.ai/production/authentication/overview#require-user-authentication) and as an "editor" role? You'll need to update the `DownloadPolicy` to avoid some policy errors. For example: ![image](https://user-images.githubusercontent.com/78053898/277842553-1d4c7f8d-7af6-4a86-b244-1078e2d03e90.png)

I added another clause to my DownloadPolicy which fixed the issue. I've also lowered the required role to VIEWER which I think makes more sense.

wangxiaoyou1993 commented on mage_ai/api/resources/DownloadResource.py:
```diff
@@ -0,0 +1,61 @@
+    def generate_download_token(file_path):
+        now = datetime.datetime.utcnow()
+        payload = {
+            "file_path": file_path,
```
nit: we prefer using single quote in the python code

wangxiaoyou1993 commented on mage_ai/server/api/downloads.py:
```diff
@@ -62,3 +66,39 @@ async def get(self, pipeline_uuid, block_uuid, **kwargs):
+            file_path = decoded_payload['file_path']
+            file_name = file_path.split('/')[-1]
+            is_temporary = file_path.startswith(TEMPORARY_DOWNLOAD_LOCATION)
+            with open(file_path, "rb") as f:
```
I feel there's a security vulnerability.

User can manually generate the token if they know the JWT_SECRET (the default value is public) and then they can download any file in the instance.

PopaRares replied:
While I believe having a default value for something as critical as a JWT secret is a security vulnerability in and of itself, we've taken measures to address this concern. At startup, we now replace the default secret with a randomly generated string value. This step effectively eliminates the risk of token forgery. However, it's important to note that this change has consequences: in the event of an environment reset, all active user sessions will become invalid.
Since the two token types are different in both scope and lifespan, we've decided to separate their secrets. The OAuth tokens will continue to use the original JWT_SECRET, while the download tokens will be generated using a new secret, JWT_DOWNLOAD_SECRET, which would get randomly generated if a specific value is not provided at startup. We're confident about this solution as it strikes a balance between security and user experience, hope you guys like it as well.

PopaRares commented on mage_ai/server/api/downloads.py:
```diff
@@ -62,3 +66,39 @@ async def get(self, pipeline_uuid, block_uuid, **kwargs):
+            decoded_payload = jwt.decode(token, JWT_SECRET, algorithms=["HS256"])
+
+            file_path = decoded_payload['file_path']
+            file_name = file_path.split('/')[-1]
```
refactored this part so this comment is no longer necessary

PopaRares commented:
Hello there,

It's been a while since I last updated this feature, but I'm thrilled to share the significant improvements I've made:

* I've created a separate JWT secret for generating download tokens. This secret is now randomly generated at startup if no value is set, addressing concerns raised by @wangxiaoyou1993.

* In the event of a pipeline download, creating the zipfile now occurs during the second call, offering several advantages:

  * The first call is faster, reducing the time the UI is locked in place.
  * This change makes the feature easily extensible for other types of downloads, such as folders or multiple files.
  * Temporary files are now created and destroyed within the same context.
* I've enhanced the process of generating [temporary files](https://docs.python.org/3/library/tempfile.html).

* My code now includes checks for relative paths that point to files outside the project's directory.

* In the first call, I've updated the response to return the download token itself instead of a pre-built URI.

* I've extracted the frontend code responsible for initiating the second download call and placed it into a utils script for better organization and maintainability.

I am eager to hear your feedback on this. Cheers!

adelcast commented on mage_ai/server/api/downloads.py:
```diff
@@ -62,3 +68,72 @@ async def get(self, pipeline_uuid, block_uuid, **kwargs):
+
+        # trying to access files outside of the project folder
+        if common_ground != self.abs_repo_path:
+            raise Exception(abs_path)
```
Can we make this Exception more specific? https://docs.python.org/3/library/exceptions.html#bltin-exceptions.. I am thinking PermissionError makes a bit more sense

PopaRares replied:
I think **PremissionError** makes it sound like there are certain permissions that would allow you to download files outside the scope of the project, which there are none. As such I replaced it with [**ValueError**](https://docs.python.org/3/library/exceptions.html#ValueError) as I think it makes the most sense.

PopaRares commented on mage_ai/api/constants.py:
```diff
@@ -52,3 +52,6 @@ class AuthorizeStatusType(str, Enum):
 META_KEY_OFFSET = '_offset'
+
+DOWNLOAD_TOKEN_LIFESPAN = 60  # Lifespan of generated download token in seconds
+TEMPORARY_DOWNLOAD_LOCATION = '/tmp'  # Temporary location of generated zipfiles
```
Addressed this.
